### ISSUE TITLE
causal mask assert hit

### ISSUE THREAD

**Issue by pfeatherstone** [lucidrains/recurrent-memory-transformer-pytorch#10]
When `causal==True` you cannot pass a `mask` in the `forward` method of `RecurrentMemoryTransformer`.
The assert statement is https://github.com/lucidrains/recurrent-memory-transformer-pytorch/blob/98bf3091a29fbd65dbbb30ce00dd1cadd05fef2d/recurrent_memory_transformer_pytorch/recurrent_memory_transformer.py#L344
Is it not possible to AND a causal mask on top of a user-defined custom mask?

**Comment by pfeatherstone** [comment]
Can we not replace:
```
assert not exists(mask)
mask = rearrange(causal_mask, 'i j -> 1 1 i j')
```
with
```
# assert not exists(mask)
causal_mask = rearrange(causal_mask, 'i j -> 1 1 i j')
if not exists(mask):
    mask = causal_mask
else:
    mask = rearrange(mask, 'b i -> b 1 i 1') & rearrange(mask, 'b i -> b 1 1 i') & causal_mask
```

**Comment by pfeatherstone** [comment]
But if set:
```
attn_bias = mask.float().masked_fill(~mask, -torch.finfo(q.dtype).max)
```
and call 
```
F.scaled_dot_product_attention(
                q, k, v,
                attn_mask = attn_bias,
                dropout_p = self.dropout if self.training else 0., 
                is_causal = self.causal
            )
```
instead then ok.
I think there might be some bugs in `F.scaled_dot_product_attention()`

**Comment by lucidrains** [comment]
@pfeatherstone why do you need masking on top of autoregressive training?

i can add this (it is tricky with the `scaled_dot_product_attention` function)

**Comment by pfeatherstone** [comment]
My data has holes/silence in it. So my mask can in theory be random

**Comment by pfeatherstone** [comment]
I'm trying to train something like TTS and STT end-to-end. But coming up with a new language. I have some losses to make the output of TTS speech-like. Otherwise, I want to train using CTC, the logits of STT and the input of TTS. I'm creating blocks of speech and silence.

**Comment by lucidrains** [comment]
how long are the sequences? can you not just use flash attention 2?

**Comment by lucidrains** [comment]
but yea, i can set aside an hour today and fix this

**Comment by pfeatherstone** [comment]
Well it looks like for RMT to work you need to train with at least 7 blocks. Each block being roughly 1024 words. But don't know. It's still not clear to me what seq_len (block size) should be. In theory even with seq_len==1 you get infinite recurrence. So don't know how long the training sequence needs to be

**Comment by lucidrains** [comment]
@pfeatherstone if your sequences are less than 16k, just stick with flash attention

if you are doing say 1 million, then yea, go with RMT

**Comment by pfeatherstone** [comment]
But in any case I need to be streaming the generated audio ad infinitum in a well definite way. RMT seemed to provide a good solution but could be wrong.

**Comment by lucidrains** [comment]
@pfeatherstone do you want to try 0.4.3 and see if it works or not?

**Comment by pfeatherstone** [comment]
Yep will do in the morning. Thank you

**Comment by lucidrains** [comment]
ok sounds good, i'll wait for you to test it again, and if it bugs out, can dig deeper

the NaNs could have to do with the fact that you masked out an entire row, but not sure

**Comment by pfeatherstone** [comment]
Yeah i get NANs.

**Comment by pfeatherstone** [comment]
Cheers that works. However, my mask wasn't completely blanked out. So I think there was some other issue that has been accidentally fixed. I'm convinced that `scaled_dot_product_attention()` has subtle bugs deep down in the C++ implementation when tensor shapes or strides aren't correct.

**Comment by lucidrains** [comment]
hmm, i'm not entirely sure then, but at least it has been fixed!

**Comment by lucidrains** [comment]
i'm going to start adding this null key/value more often, given [this new research](https://arxiv.org/abs/2306.12929)

**Comment by pfeatherstone** [comment]
Haven't you done this before with [learned tokens](https://github.com/lucidrains/x-transformers#memory-transformers) ? Feels like it's achieving the same thing no ?

**Comment by pfeatherstone** [comment]
Yeah sure, but i mean having the learned tokens there means you can attend to them, not the input tokens, and essentially attend to nothing.

**Comment by lucidrains** [comment]
@pfeatherstone yeah, you could always add some fixed 'read memory tokens' (as in the terminology of this paper)

you are right
